pkg-r/R/contents_shinychat.R

```diff
@@ -1,14 +1,261 @@
+  content
+) {
+  props <- list(
+    "request-id" = content@id,
+    name = content@name,
+    arguments = jsonlite::toJSON(content@arguments, auto_unbox = TRUE)
+  )
+
+  # Add optional title if present in tool annotations
+  tool <- content@tool
+  if (!is.null(tool) && !is.null(tool@annotations$title)) {
+    props$title <- content@tool@annotations$title
+  }
+
+  # Add optional intent if present in request arguments
+  if (!is.null(content@arguments$.intent)) {
+    props$intent <- content@arguments$.intent
+  }
+
+  # Return structured tag
+  htmltools::tag("shiny-tool-request", props)
+}
+
+S7::method(contents_shinychat, ellmer::ContentToolResult) <- function(content) {
+  # Prepare base props
+  props <- list(
+    request_id = content@request@id,
+    request_call = "",
+    name = content@request@name,
+    status = if (tool_errored(content)) "error" else "success",
+    show_request = tolower(!isFALSE(content@extra$display_tool_request))
+  )
+
+  icon_deps <- NULL
+  tool <- content@request@tool
+
+  if (!is.null(tool)) {
+    # Format fails if tool is not present (ellmer v0.3.0, tidyverse/ellmer#691)
+    props$request_call <- format(content@request, show = "call")
+
+    # Add optional title if present
+    if (!is.null(tool@annotations$title)) {
+      props$title <- tool@annotations$title
+    }
+
+    # Add optional icon if present
+    if (!is.null(tool@annotations$icon)) {
+      props$icon <- tool@annotations$icon
+      icon_deps <- htmltools::findDependencies(props$icon)
+    }
+  } else {
+    props$request_call <- jsonlite::toJSON(
+      list(
+        id = content@request@id,
+        name = content@request@name,
+        arguments = content@request@arguments
+      ),
+      auto_unbox = TRUE,
+      pretty = 2
+    )
+  }
+
+  # Add optional intent if present
+  if (!is.null(content@request@arguments$.intent)) {
+    props$intent <- content@request@arguments$.intent
+  }
+
+  display_props <- tool_result_display(content)
+  props$value <- display_props$value
+  props$value_type <- display_props$value_type
+
+  names(props) <- gsub("_", "-", names(props))
+
+  htmltools::tag(
+    "shiny-tool-result",
+    compact(list2(!!!props, display_props$deps, icon_deps))
+  )
+}
+
+tool_result_display <- function(content) {
+  # Default fallback
+  res <- list(
+    value = tool_string(content),
+    value_type = "code",
+    deps = NULL
+  )
+
+  if (tool_errored(content)) {
+    return(res)
+  }
+
+  # Get display from extra data attached to the result
+  display <- content@extra$display
+
+  # If no display, return code value
+  if (is.null(display)) {
+    return(res)
+  }
+
+  is_html <-
+    inherits(display, c("html", "shiny.tag", "shiny.tag.list")) ||
+    (is.list(display) && !is.null(display$html))
+
+  if (is_html) {
+    html <- list(
+      value = format(display),
+      value_type = "html",
+      deps = htmltools::findDependencies(display)
+    )
+    return(html)
```

Comment:
I think you could get away with just returning `value` and `value_type` from this function?

```suggestion
    return(list(value = display, value_type = "html"))
```

Reply:
There's some overlapping logic between `value_type` and `value` that doesn't make it worth separating into two functions. But I was able to clean this up quite a bit with a second pass in f3c28fa85b2e226c964793604e6d6b3b54809c9f

Reply:
Nice, thanks. 👍